In [1]:
# Adding autoreload will let you edit your package and see the changes without restarting the notebook.

%load_ext autoreload

%autoreload 2

In [2]:
import pandas as pd

In [3]:
deliveries = pd.read_json("../data/raw/deliveries.jsonl", lines=True)
products = pd.read_json("../data/raw/products.jsonl", lines=True)
sessions = pd.read_json("../data/raw/sessions.jsonl", lines=True)
users = pd.read_json("../data/raw/users.jsonl", lines=True)

### Analiza jakosci danych w zbiorze "deliveries"

In [8]:
deliveries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2975 entries, 0 to 2974
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   purchase_id         2975 non-null   int64         
 1   purchase_timestamp  2975 non-null   datetime64[ns]
 2   delivery_timestamp  2823 non-null   datetime64[ns]
 3   delivery_company    2844 non-null   Int64         
dtypes: Int64(1), datetime64[ns](2), int64(1)
memory usage: 96.0 KB


In [12]:
deliveries

purchase_id  purchase_timestamp  delivery_timestamp  delivery_company
0           20001 2021-01-09 20:27:33 2021-01-11 20:28:09               620
1           20002 2021-01-09 16:11:11 2021-01-12 16:13:44               360
2           20003 2021-01-09 20:33:15 2021-01-11 20:37:51               620
3           20004 2021-01-08 05:23:04 2021-01-12 05:23:36               516
4           20005 2021-01-10 19:50:56 2021-01-12 19:53:31               620
...           ...                 ...                 ...               ...
2970        22971 2021-01-03 02:05:42 2021-01-08 02:09:59               620
2971        22972 2021-01-02 22:00:31 2021-01-06 22:00:51               360
2972        22973 2021-01-01 21:30:29 2021-01-05 21:31:01               620
2973        22974 2021-01-03 04:38:40                 NaT               360
2974        22975 2021-01-01 13:05:38 2021-01-03 13:07:44              <NA>

[2975 rows x 4 columns]

In [21]:
deliveries['purchase_timestamp'].agg(['min', 'max'])

min   2021-01-01 03:23:09
max   2021-10-27 14:54:44
Name: purchase_timestamp, dtype: datetime64[ns]

Rozumiem, że deliveries 

### Analiza jakości danych w zbiorze "products"

In [10]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 319 entries, 0 to 318
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     319 non-null    int64  
 1   product_name   319 non-null    object 
 2   category_path  319 non-null    object 
 3   price          319 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 10.1+ KB


In [11]:
products

product_id                                    product_name  \
0          1001                   Telefon Siemens Gigaset DA310   
1          1002                              Kyocera FS-1135MFP   
2          1003                              Kyocera FS-3640MFP   
3          1004                            Fallout 3 (Xbox 360)   
4          1005  Szalone Króliki Na żywo i w kolorze (Xbox 360)   
..          ...                                             ...   
314        1315                                      Jabra Talk   
315        1316                      Plantronics Voyager Legend   
316        1317                           Plantronics Savi W740   
317        1318                           Plantronics Savi W710   
318        1319                                     HTC HS-S200   

                                         category_path        price  
0            Telefony i akcesoria;Telefony stacjonarne        58.97  
1    Komputery;Drukarki i skanery;Biurowe urządzeni...      2048.50  
2    Komputery;Drukarki i skanery;Biurowe urządzeni...      7639.00  
3            Gry i konsole;Gry na konsole;Gry Xbox 360  49990000.00  
4            Gry i konsole;Gry na konsole;Gry Xbox 360       -49.99  
..                                                 ...          ...  
314  Telefony i akcesoria;Akcesoria telefoniczne;Ze...        54.99  
315  Telefony i akcesoria;Akcesoria telefoniczne;Ze...       249.00  
316  Telefony i akcesoria;Akcesoria telefoniczne;Ze...      1303.96  
317                         Sprzęt RTV;Audio;Słuchawki      -553.00  
318  Telefony i akcesoria;Akcesoria telefoniczne;Ze...        14.99  

[319 rows x 4 columns]

In [24]:
products[products['price'] < 0]

product_id                                       product_name  \
4          1005     Szalone Króliki Na żywo i w kolorze (Xbox 360)   
37         1038                                    LCD Asus VK278Q   
46         1047  Pirates of the Caribbean At World&#39;s End (PS3)   
51         1052                            Duke Nukem Forever (PC)   
54         1055                 Call of Duty Modern Warfare 3 (PC)   
129        1130  Tom Clancy&#39;s Ghost Recon 4 Future Soldier ...   
153        1154                                Civilization 4 (PC)   
163        1164                       Aliens Colonial Marines (PC)   
201        1202                              Final Fantasy 14 (PC)   
202        1203                 LEGO Batman 2 DC Super Heroes (PC)   
211        1212                                        Arma 3 (PC)   
220        1221                   The Sims 3 Cztery Pory Roku (PC)   
221        1222                                 Parrot Minikit NEO   
225        1226                             Supreme Commander (PC)   
233        1234                                     Sony DVP-SR760   
251        1252                                  Two Worlds 2 (PC)   
254        1255                                          Auta (PC)   
259        1260                                 Sleeping Dogs (PC)   
295        1296                                     Sencor SDA-630   
317        1318                              Plantronics Savi W710   

                                         category_path    price  
4            Gry i konsole;Gry na konsole;Gry Xbox 360   -49.99  
37                     Komputery;Monitory;Monitory LCD -1117.01  
46       Gry i konsole;Gry na konsole;Gry PlayStation3   -89.90  
51                       Gry i konsole;Gry komputerowe   -78.90  
54                       Gry i konsole;Gry komputerowe   -32.99  
129          Gry i konsole;Gry na konsole;Gry Xbox 360   -49.99  
153                      Gry i konsole;Gry komputerowe   -40.90  
163                      Gry i konsole;Gry komputerowe  -112.99  
201                      Gry i konsole;Gry komputerowe  -135.90  
202                      Gry i konsole;Gry komputerowe   -29.99  
211                      Gry i konsole;Gry komputerowe  -102.90  
220                      Gry i konsole;Gry komputerowe   -31.90  
221  Telefony i akcesoria;Akcesoria telefoniczne;Ze...  -245.00  
225                      Gry i konsole;Gry komputerowe   -41.90  
233                   Sprzęt RTV;Video;Odtwarzacze DVD  -193.00  
251                      Gry i konsole;Gry komputerowe   -12.90  
254                      Gry i konsole;Gry komputerowe   -32.90  
259                      Gry i konsole;Gry komputerowe   -74.90  
295  Sprzęt RTV;Video;Telewizory i akcesoria;Anteny...   -89.99  
317                         Sprzęt RTV;Audio;Słuchawki  -553.00

In [ ]:
products[products['price'] < 0].shape[0]

In [46]:
products[products['price'] > 5000].sort_values(by='price')

product_id                                      product_name  \
80         1081                                        OKI B840dn   
79         1080                                Kyocera FS-3140MFP   
2          1003                                Kyocera FS-3640MFP   
111        1112                                  Singularity (PC)   
150        1151                                      Mafia 2 (PC)   
139        1140                       Euro Truck Simulator 2 (PC)   
196        1197                             Hearts of Iron 3 (PC)   
213        1214     Call of Cthulhu Mroczne zakamarki Świata (PC)   
216        1217                          Hitman Krwawa Forsa (PC)   
268        1269                                   Space Hack (PC)   
235        1236                    Avencast Rise of the Mage (PC)   
245        1246                                      Quake 4 (PC)   
218        1219                                   Overlord 2 (PC)   
181        1182                                Battlefield 3 (PC)   
3          1004                              Fallout 3 (Xbox 360)   
253        1254  Warhammer 40.000 Dawn of War Winter Assault (PC)   
305        1306                               One For All SV 9215   
128        1129              Battlefield Bad Company 2 (Xbox 360)   
123        1124                              Devil May Cry 4 (PC)   
200        1201                                   Jabra Speak 410   
78         1079                                Kyocera FS-3040MFP   
77         1078                                Kyocera FS-3540MFP   

                                         category_path         price  
80   Komputery;Drukarki i skanery;Biurowe urządzeni...  5.259000e+03  
79   Komputery;Drukarki i skanery;Biurowe urządzeni...  5.301900e+03  
2    Komputery;Drukarki i skanery;Biurowe urządzeni...  7.639000e+03  
111                      Gry i konsole;Gry komputerowe  8.900000e+06  
150                      Gry i konsole;Gry komputerowe  1.190000e+07  
139                      Gry i konsole;Gry komputerowe  1.299000e+07  
196                      Gry i konsole;Gry komputerowe  1.599000e+07  
213                      Gry i konsole;Gry komputerowe  1.899000e+07  
216                      Gry i konsole;Gry komputerowe  1.990000e+07  
268                      Gry i konsole;Gry komputerowe  1.999000e+07  
235                      Gry i konsole;Gry komputerowe  3.199000e+07  
245                      Gry i konsole;Gry komputerowe  3.399000e+07  
218                      Gry i konsole;Gry komputerowe  3.499000e+07  
181                      Gry i konsole;Gry komputerowe  3.690000e+07  
3            Gry i konsole;Gry na konsole;Gry Xbox 360  4.999000e+07  
253                      Gry i konsole;Gry komputerowe  5.590000e+07  
305  Sprzęt RTV;Video;Telewizory i akcesoria;Anteny...  7.999000e+07  
128          Gry i konsole;Gry na konsole;Gry Xbox 360  8.999000e+07  
123                      Gry i konsole;Gry komputerowe  9.400000e+07  
200  Telefony i akcesoria;Akcesoria telefoniczne;Ze...  3.274700e+08  
78   Komputery;Drukarki i skanery;Biurowe urządzeni...  4.598000e+09  
77   Komputery;Drukarki i skanery;Biurowe urządzeni...  5.126000e+09

In [36]:
products[products['price'] > 5000].shape[0]

22

In [42]:
products['category_path'].value_counts()

Gry i konsole;Gry komputerowe                                        202
Gry i konsole;Gry na konsole;Gry Xbox 360                             32
Sprzęt RTV;Video;Telewizory i akcesoria;Anteny RTV                    30
Komputery;Monitory;Monitory LCD                                       17
Komputery;Drukarki i skanery;Biurowe urządzenia wielofunkcyjne         9
Gry i konsole;Gry na konsole;Gry PlayStation3                          9
Telefony i akcesoria;Akcesoria telefoniczne;Zestawy głośnomówiące      5
Telefony i akcesoria;Akcesoria telefoniczne;Zestawy słuchawkowe        4
Telefony i akcesoria;Telefony komórkowe                                2
Sprzęt RTV;Video;Odtwarzacze DVD                                       2
Komputery;Tablety i akcesoria;Tablety                                  2
Sprzęt RTV;Przenośne audio i video;Odtwarzacze mp3 i mp4               2
Telefony i akcesoria;Telefony stacjonarne                              1
Sprzęt RTV;Video;Telewizory i akcesoria;Okulary 3D 

### Analiza jakości danych w zbiorze "sessions"

In [13]:
sessions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 173632 entries, 0 to 173631
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   session_id        173632 non-null  int64         
 1   timestamp         173632 non-null  datetime64[ns]
 2   user_id           164909 non-null  Int64         
 3   product_id        165137 non-null  Int64         
 4   event_type        173632 non-null  object        
 5   offered_discount  173632 non-null  int64         
 6   purchase_id       2975 non-null    Int64         
dtypes: Int64(3), datetime64[ns](1), int64(2), object(1)
memory usage: 9.8+ MB


In [14]:
sessions

session_id           timestamp  user_id  product_id    event_type  \
0           100001 2021-01-09 20:25:15      104        1044  VIEW_PRODUCT   
1           100001 2021-01-09 20:27:03      104        1043  VIEW_PRODUCT   
2           100001 2021-01-09 20:27:33      104        1043   BUY_PRODUCT   
3           100002 2021-01-09 15:07:56      107        1305  VIEW_PRODUCT   
4           100002 2021-01-09 15:08:11      107        1292  VIEW_PRODUCT   
...            ...                 ...      ...         ...           ...   
173627      109864 2021-01-01 12:55:09      297        1064  VIEW_PRODUCT   
173628      109864 2021-01-01 12:56:26      297        1033  VIEW_PRODUCT   
173629      109864 2021-01-01 13:00:56      297        1017  VIEW_PRODUCT   
173630      109864 2021-01-01 13:05:38      297        1017   BUY_PRODUCT   
173631      109865 2021-01-03 14:38:26      301        1064  VIEW_PRODUCT   

        offered_discount  purchase_id  
0                      0         <NA>  
1                      0         <NA>  
2                      0        20001  
3                     10         <NA>  
4                     10         <NA>  
...                  ...          ...  
173627                10         <NA>  
173628                10         <NA>  
173629                10         <NA>  
173630                10        22975  
173631                 5         <NA>  

[173632 rows x 7 columns]

In [32]:
sessions['timestamp'].agg(['min', 'max'])

min   2021-01-01 00:00:22
max   2021-10-27 15:37:51
Name: timestamp, dtype: datetime64[ns]

In [18]:
sessions[['event_type']].value_counts()

event_type  
VIEW_PRODUCT    170657
BUY_PRODUCT       2975
dtype: int64

My nie zrobimy nic jak mamy tylko 3000 kupionych produktów

In [27]:
buy_sessions = sessions[sessions['event_type'] == 'BUY_PRODUCT']

In [29]:
buy_sessions.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2975 entries, 2 to 173630
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   session_id        2975 non-null   int64         
 1   timestamp         2975 non-null   datetime64[ns]
 2   user_id           2822 non-null   Int64         
 3   product_id        2826 non-null   Int64         
 4   event_type        2975 non-null   object        
 5   offered_discount  2975 non-null   int64         
 6   purchase_id       2975 non-null   Int64         
dtypes: Int64(3), datetime64[ns](1), int64(2), object(1)
memory usage: 194.7+ KB


- 153 sesji bez user_id - ale user_id nie jest dla nas krytyczny
- 149 sesji abez product_id - nie możemy użyć tych sesji bez product_ids

### Check if there are purchases in deliveries whic hare not in sessions and vice verca

In [43]:
deliveries[~deliveries['purchase_id'].isin(sessions['purchase_id'])]

Empty DataFrame
Columns: [purchase_id, purchase_timestamp, delivery_timestamp, delivery_company]
Index: []

In [45]:
sessions[~sessions['purchase_id'].isin(deliveries['purchase_id'])]['event_type'].value_counts()

VIEW_PRODUCT    170657
Name: event_type, dtype: int64

### Analiza jakości danych w zbiorze "users"

In [19]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  200 non-null    int64 
 1   name     200 non-null    object
 2   city     200 non-null    object
 3   street   200 non-null    object
dtypes: int64(1), object(3)
memory usage: 6.4+ KB


In [20]:
users

user_id                  name      city                  street
0        102       Tymoteusz Wołos  Warszawa          al. Żwirowa 08
1        103      Olgierd Wardziak    Kraków  aleja Wybickiego 12/30
2        104       Urszula Kowaluk    Kraków  ulica Strumykowa 51/91
3        105  Roksana Hajdukiewicz  Szczecin      ul. Wyzwolenia 679
4        106         Julian Fulara   Wrocław        aleja Dworska 63
..       ...                   ...       ...                     ...
195      297      Marcelina Guziak  Warszawa  plac Słowackiego 43/14
196      298       Marianna Koźlak     Radom      aleja Stroma 64/71
197      299        Tymoteusz Tórz    Kraków      plac Matejki 24/06
198      300         Łukasz Faryna    Kraków       al. Zakątek 36/81
199      301       Ryszard Trzmiel     Radom     plac Kasprowicza 31

[200 rows x 4 columns]